### ISSUE TITLE
[Bug]: Vision encoder attention outputs uniform-average with npu_fused_infer_attention_score (TND + sparse_mode=0 + no mask)

### ISSUE THREAD

**Issue by shijunan126-creator** [vllm-project/vllm-ascend#13699]
# [Bug]: Vision encoder attention outputs uniform-average with `npu_fused_infer_attention_score` (TND + sparse_mode=0 + no mask)

> 建议提交到: vllm-project/vllm-ascend
> 类型: Bug(静默数值错误,非崩溃——图片请求"能跑但注意力失效")

## Summary

`AscendMMEncoderAttention` runs vision-encoder attention via
`torch_npu.npu_fused_infer_attention_score` with `input_layout="TND"`,
`sparse_mode=0` and **no attention mask**. On this stack the operator returns a
**uniform-average output** (softmax is effectively not applied): error vs an
fp32 reference is ~3.1 (should be ~3e-3 for fp16), for both single-sequence and
multi-sequence (multi-image) inputs. The LLM (decode) paths are unaffected
because they use `sparse_mode=3` + explicit mask or paged mode.

## Your current environment

<details>
<summary>collect_env</summary>

```text
PyTorch version: 2.10.0+cpu
Is debug build: False

OS: Ubuntu 22.04.5 LTS (aarch64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: 15.0.7
CMake version: version 4.4.0
Libc version: glibc-2.35

Python version: 3.12.13 (main, Jul  7 2026, 16:29:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.10.0-216.0.0.115.oe2203sp4.aarch64-aarch64-with-glibc2.35

CPU:
Architecture:                       aarch64
CPU op-mode(s):                     64-bit
Byte Order:                         Little Endian
CPU(s):                             192
On-line CPU(s) list:                0-191
Vendor ID:                          HiSilicon
BIOS Vendor ID:                     HiSilicon
Model name:                         Kunpeng-920
BIOS Model name:                    HUAWEI Kunpeng 920 5250
Model:                              0
Thread(s) per core:                 1
Core(s) per socket:                 48
Socket(s):                          4
Stepping:                           0x1
Frequency boost:                    disabled
CPU max MHz:                        2600.0000
CPU min MHz:                        200.0000
BogoMIPS:                           200.00
Flags:                              fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma dcpop asimddp asimdfhm ssbs
L1d cache:                          12 MiB (192 instances)
L1i cache:                          12 MiB (192 instances)
L2 cache:                           96 MiB (192 instances)
L3 cache:                           192 MiB (8 instances)
NUMA node(s):                       8
NUMA node0 CPU(s):                  0-23
NUMA node1 CPU(s):                  24-47
NUMA node2 CPU(s):                  48-71
NUMA node3 CPU(s):                  72-95
NUMA node4 CPU(s):                  96-119
NUMA node5 CPU(s):                  120-143
NUMA node6 CPU(s):                  144-167
NUMA node7 CPU(s):                  168-191
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; __user pointer sanitization
Vulnerability Spectre v2:           Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pyzmq==27.1.0
[pip3] torch==2.10.0+cpu
[pip3] torch_npu==2.10.0.post2
[pip3] torchaudio==2.10.0+cpu
[pip3] torchvision==0.25.0+cpu
[pip3] transformers==5.5.4
[conda] Could not collect
vLLM Version: 0.23.0
vLLM Ascend Version: 0.23.0rc1

ENV Variables:
ASCEND_TOOLKIT_LATEST_HOME=/usr/local/Ascend/ascend-toolkit/latest
ATB_OPSRUNNER_KERNEL_CACHE_LOCAL_COUNT=1
ATB_STREAM_SYNC_EVERY_RUNNER_ENABLE=0
ATB_STREAM_SYNC_EVERY_KERNEL_ENABLE=0
ATB_CXX_ABI=1
ATB_OPSRUNNER_KERNEL_CACHE_GLOABL_COUNT=5
ATB_HOME_PATH=/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_1
ASCEND_TOOLKIT_HOME=/usr/local/Ascend/cann-9.0.1
ATB_COMPARE_TILING_EVERY_KERNEL=0
ASCEND_OPP_PATH=/usr/local/Ascend/cann-9.0.1/opp
LD_LIBRARY_PATH=/usr/local/Ascend/cann-9.0.1/lib64:/usr/local/Ascend/cann-9.0.1/lib64/plugin/opskernel:/usr/local/Ascend/cann-9.0.1/lib64/plugin/nnengine:/usr/local/Ascend/cann-9.0.1/opp/built-in/op_impl/ai_core/tbe/op_tiling:/usr/local/Ascend/driver/lib64:/usr/local/Ascend/driver/lib64/common:/usr/local/Ascend/driver/lib64/driver:/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_1/lib:/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_1/examples:/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_1/tests/atbopstest:/usr/local/Ascend/ascend-toolkit/latest/tools/aml/lib64:/usr/local/Ascend/ascend-toolkit/latest/tools/aml/lib64/plugin:/usr/local/Ascend/ascend-toolkit/latest/lib64:/usr/local/Ascend/ascend-toolkit/latest/lib64/plugin/opskernel:/usr/local/Ascend/ascend-toolkit/latest/lib64/plugin/nnengine:/usr/local/Ascend/ascend-toolkit/latest/opp/built-in/op_impl/ai_core/tbe/op_tiling:/usr/local/python3.12.13/lib::/usr/local/lib
ASCEND_AICPU_PATH=/usr/local/Ascend/cann-9.0.1
ATB_STREAM_SYNC_EVERY_OPERATION_ENABLE=0
ASCEND_HOME_PATH=/usr/local/Ascend/cann-9.0.1
ATB_MATMUL_SHUFFLE_K_ENABLE=1
ATB_WORKSPACE_MEM_ALLOC_ALG_TYPE=1
ATB_SHARE_MEMORY_NAME_SUFFIX=
TORCH_DEVICE_BACKEND_AUTOLOAD=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_root
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
```

```text
NPU:
+------------------------------------------------------------------------------------------------+
| npu-smi 25.5.2                   Version: 25.5.2                                               |
+---------------------------+---------------+----------------------------------------------------+
| NPU   Name                | Health        | Power(W)    Temp(C)           Hugepages-Usage(page)|
| Chip                      | Bus-Id        | AICore(%)   Memory-Usage(MB)  HBM-Usage(MB)        |
+===========================+===============+====================================================+
| 0     910B4-1             | OK            | 95.5        42                0    / 0             |
| 0                         | 0000:C1:00.0  | 0           0    / 0          60735/ 65536         |
+===========================+===============+====================================================+
| 1     910B4-1             | OK            | 93.4        38                0    / 0             |
| 0                         | 0000:C2:00.0  | 0           0    / 0          60731/ 65536         |
+===========================+===============+====================================================+
| 2     910B4-1             | OK            | 88.1        36                0    / 0             |
| 0                         | 0000:81:00.0  | 0           0    / 0          3415 / 65536         |
+===========================+===============+====================================================+
| 3     910B4-1             | OK            | 93.4        38                0    / 0             |
| 0                         | 0000:82:00.0  | 0           0    / 0          3418 / 65536         |
+===========================+===============+====================================================+
| 4     910B4-1             | OK            | 90.9        45                0    / 0             |
| 0                         | 0000:01:00.0  | 0           0    / 0          3411 / 65536         |
+===========================+===============+====================================================+
| 5     910B4-1             | OK            | 90.7        43                0    / 0             |
| 0                         | 0000:02:00.0  | 0           0    / 0          3416 / 65536         |
+===========================+===============+====================================================+
| 6     910B4-1             | OK            | 99.2        41                0    / 0             |
| 0                         | 0000:41:00.0  | 0           0    / 0          3418 / 65536         |
+===========================+===============+====================================================+
| 7     910B4-1             | OK            | 87.6        41                0    / 0             |
| 0                         | 0000:42:00.0  | 0           0    / 0          3416 / 65536         |
+===========================+===============+====================================================+
+---------------------------+---------------+----------------------------------------------------+
| NPU     Chip              | Process id    | Process name             | Process memory(MB)      |
+===========================+===============+====================================================+
| 0       0                 | 4049849       |                          | 57364                   |
+===========================+===============+====================================================+
| 1       0                 | 4050012       |                          | 57364                   |
+===========================+===============+====================================================+
| No running processes found in NPU 2                                                            |
+===========================+===============+====================================================+
| No running processes found in NPU 3                                                            |
+===========================+===============+====================================================+
| No running processes found in NPU 4                                                            |
+===========================+===============+====================================================+
| No running processes found in NPU 5                                                            |
+===========================+===============+====================================================+
| No running processes found in NPU 6                                                            |
+===========================+===============+====================================================+
| No running processes found in NPU 7                                                            |
+===========================+===============+====================================================+

CANN:
package_name=Ascend-cann-toolkit
version=9.0.1
innerversion=V100R001C10SPC002B220
compatible_version=[V100R001C21],[V100R001C22],[V100R001C23],[V100R001C25],[V100R001C10]
arch=aarch64
os=linux
path=/usr/local/Ascend/cann-9.0.1
```
</details>

## 🐛 Describe the bug

Vision requests complete without errors, but the attention output of every ViT
layer degenerates to the per-token **mean of V** (uniform softmax). Reproduced
with a minimal standalone script (no model load):

### Reproduction (single sequence, T=2048)

```python
import torch, torch_npu
torch.npu.set_device(0)
H, D, N = 16, 128, 2048
scale = D ** -0.5
q = torch.randn(N, H, D, dtype=torch.float16, device="npu")
k = torch.randn(N, H, D, dtype=torch.float16, device="npu")
v = torch.randn(N, H, D, dtype=torch.float16, device="npu")

# fp32 reference
qf, kf, vf = q.float(), k.float(), v.float()
ref = torch.softmax((qf @ kf.transpose(-2, -1)) * scale, dim=-1) @ vf

out, _ = torch_npu.npu_fused_infer_attention_score(
    query=q, key=k, value=v, atten_mask=None, block_table=None,
    input_layout="TND", block_size=128,
    actual_seq_lengths=torch.tensor([N], dtype=torch.int64),
    actual_seq_lengths_kv=torch.tensor([N], dtype=torch.int64),
    num_key_value_heads=H, num_heads=H, scale=scale,
    pre_tokens=2147483647, next_tokens=2147483647, sparse_mode=0)

err = (out.float() - ref).abs().max().item()
print("err:", err)          # 实测 ~3.1 (fp16 正常水平应 ~3e-3)
print("out mean:", out.float().abs().mean().item())   # ~0.03 ≈ mean(|v|) → 均匀平均
```

### Measured results

| config | err vs fp32 ref | note |
|---|---|---|
| TND, sm=0, no mask, single seq (2048) | **3.1** | out ≈ uniform average |
| TND, sm=0, no mask, 2×1024 seqs | **3.1** | same |
| sm=3 + full-True bool mask | ~2.8 | also wrong |
| sm=3 + full-False bool mask | ~2.9 | also wrong |
| sm=3 + causal mask (LLM path) | correct | LLM unaffected |
| native matmul+softmax+matmul | **0.003** | correct fp16 |

Attempted parameter variants all produce the same wrong output: `int64`/`int32`
tensors vs `list[int]` seq lengths, `block_size` 0/128, `pre_tokens` 65535/INT_MAX,
mask dtypes bool/int8, transposed layouts, bf16. `scale` does change the output
magnitude (so it is applied), but no parameter combination matches the reference.

## Root cause analysis

1. **`MaskType::NO_MASK` is a decode-only path.** In
   `fused_infer_attention_score_tiling.cpp` the only registered tiling class for
   NO_MASK is annotated `// Decoding 场景 (pagedCacheFlag == true && qSeqlen == 1
   && NO_MASK && !lseFlag)`. "No mask" in FIA is not a general full-attention
   concept; it is the decode fast path where softmax is evaluated over the
   paged KV blocks.
2. **The vision-encoder combination is untested.** The operator test suite
   (`tests/tiny_rdv/case_json`) contains no case for **D=128 + TND + non-paged +
   sparse_mode=0 + no mask**: new-template TND cases are all paged; the only
   non-paged TND case uses d=64 (old template). The encoder runs exactly this
   uncovered combination.
3. **Only the vision encoder uses it.** In vllm-ascend,
   `AscendMMEncoderAttention._run_vit_fia` is the only caller of
   "non-paged + long-seq + no mask + TND", so LLM serving (masked or paged
   paths) hides the defect — image features are silently degraded while text
   inference looks normal.

## Impact

- All vision-encoder attention in vllm-ascend eager mode outputs uniform-pooled
  features (softmax not applied) — image understanding quality is degraded
  while requests still "work".
- Performance of FIA in this mode is also poor (~311 µs kernel vs ~160 µs for
  the native chain), consistent with the operator running an unintended path.

## Performance impact of the workaround

Benchmark: 64 requests / 32 concurrency, 1080P random image per request
(`--dataset-name random-mm`, no mmcache hits), 128 in / 128 out tokens,
w8a8 quantized, qwen3.6-flash on 2×910B4 (TP2/EP):

| config | req/s | median TTFT (ms) | P99 TTFT (ms) | failed |
|---|---|---|---|---|
| eager + FIA (baseline) | 1.99 | 6220 | 16482 | 0 |
| encoder graph + FIA | — | — | — | crashes (TensorMove) on replay, no data |
| eager + native (this fix) | 2.97 | 4336 | 7644 | 0 |
| encoder graph + native | 2.95 | 4455 | 7585 | 0 |

- Throughput +49%, TTFT −30~54% vs the eager+FIA baseline.
- Single-op: FIA ~311 µs (profiler kernel dur) vs native chain ~110 µs (~3x);
  simulated ViT layer (identical weights): 1.479 ms → 1.051 ms (−29%/layer).
- The native chain also removes the encoder-graph replay crash (TensorMove,
  program id 75); with native attention the encoder ACL graphs capture and
  replay correctly (9 budget graphs), though CPU is not the bottleneck on this
  192-core host, so graph mode adds no throughput on top of the native fix.

**Comment by yiz-liu** [comment]
I think this should be assigned to @keyi-zz , please take a look.
